File: js/src/interiorPointLine.ts
import type { Geometry, Position } from "geojson";

/**
 * Computes an interior point of a line geometry (LineString/MultiLineString).
 *
 * Algorithm (ported from JTS InteriorPointLine.java):
 * 1. Compute the centroid of all linear components.
 * 2. Find the interior vertex (not an endpoint) closest to the centroid.
 * 3. If no interior vertices exist, fall back to the closest endpoint.
 *
 * @param geometry - A GeoJSON geometry (LineString, MultiLineString, or GeometryCollection containing lines)
 * @returns A position [x, y] on the geometry, or null if the geometry has no linear components
 */
export function interiorPointLine(geometry: Geometry): Position | null {
  const lines = collectLineCoords(geometry);
  if (lines.length === 0) return null;

  const centroid = computeLineCentroid(lines);

  // Phase 1: try interior vertices (indices 1..n-2)
  let best: Position | null = null;
  let bestDist = Infinity;

  for (const coords of lines) {
    for (let i = 1; i < coords.length - 1; i++) {
      const dist = distanceSq(coords[i], centroid);
      if (dist < bestDist) {
        bestDist = dist;
        best = coords[i];
      }
    }
  }

  // Phase 2: fall back to endpoints if no interior vertices found
  if (best === null) {
    for (const coords of lines) {
      for (const p of [coords[0], coords[coords.length - 1]]) {
        const dist = distanceSq(p, centroid);
        if (dist < bestDist) {
          bestDist = dist;
          best = p;
        }
      }
    }
  }

  return best;
}

/** Recursively collect coordinate arrays from all linear components. */
function collectLineCoords(geometry: Geometry): Position[][] {
  switch (geometry.type) {
    case "LineString":
      return geometry.coordinates.length > 0 ? [geometry.coordinates] : [];
    case "MultiLineString":
      return geometry.coordinates.filter((c) => c.length > 0);
    case "GeometryCollection":
      return geometry.geometries.flatMap(collectLineCoords);
    default:
      return [];
  }
}

/**
 * Compute the length-weighted centroid of line segments.
 * Each segment's midpoint is weighted by its length.
 */
function computeLineCentroid(lines: Position[][]): Position {
  let totalLen = 0;
  let cx = 0;
  let cy = 0;

  for (const coords of lines) {
    for (let i = 0; i < coords.length - 1; i++) {
      const dx = coords[i + 1][0] - coords[i][0];
      const dy = coords[i + 1][1] - coords[i][1];
      const len = Math.sqrt(dx * dx + dy * dy);
      totalLen += len;
      cx += (len * (coords[i][0] + coords[i + 1][0])) / 2;
      cy += (len * (coords[i][1] + coords[i + 1][1])) / 2;
    }
  }

  if (totalLen === 0) {
    // Degenerate: all zero-length segments — use first point
    return lines[0][0];
  }
  return [cx / totalLen, cy / totalLen];
}

function distanceSq(a: Position, b: Position): number {
  const dx = a[0] - b[0];
  const dy = a[1] - b[1];
  return dx * dx + dy * dy;
}

File: rs/core/src/interior_point_line.rs
//! Interior point computation for line geometries (LineString/MultiLineString).
//!
//! Finds the vertex closest to the centroid.
//! Ported from JTS InteriorPointLine.java.

use geo_types::{Coord, Geometry, LineString};

/// Computes an interior point of line geometries within the given geometry.
///
/// Algorithm:
/// 1. Compute the length-weighted centroid of all linear components.
/// 2. Find the interior vertex (not an endpoint) closest to the centroid.
/// 3. If no interior vertices exist, fall back to the closest endpoint.
///
/// Returns `None` if the geometry contains no line segments.
pub(crate) fn interior_point_line(geometry: &Geometry<f64>) -> Option<Coord<f64>> {
    let lines = collect_lines(geometry);
    if lines.is_empty() {
        return None;
    }

    let centroid = compute_line_centroid(&lines);

    // Phase 1: try interior vertices (indices 1..n-1, exclusive of endpoints)
    let mut best: Option<Coord<f64>> = None;
    let mut best_dist = f64::INFINITY;

    for line in &lines {
        let coords: Vec<Coord<f64>> = line.coords().copied().collect();
        for coord in coords.iter().skip(1).take(coords.len().saturating_sub(2)) {
            let dist = distance_sq(*coord, centroid);
            if dist < best_dist {
                best_dist = dist;
                best = Some(*coord);
            }
        }
    }

    // Phase 2: fall back to endpoints if no interior vertices found
    if best.is_none() {
        for line in &lines {
            let coords: Vec<Coord<f64>> = line.coords().copied().collect();
            if coords.is_empty() {
                continue;
            }
            for &p in &[coords[0], coords[coords.len() - 1]] {
                let dist = distance_sq(p, centroid);
                if dist < best_dist {
                    best_dist = dist;
                    best = Some(p);
                }
            }
        }
    }

    best
}

/// Recursively collect all LineStrings from the geometry.
fn collect_lines(geometry: &Geometry<f64>) -> Vec<&LineString<f64>> {
    match geometry {
        Geometry::LineString(ls) => {
            if ls.0.is_empty() {
                vec![]
            } else {
                vec![ls]
            }
        }
        Geometry::MultiLineString(mls) => mls.0.iter().filter(|ls| !ls.0.is_empty()).collect(),
        Geometry::GeometryCollection(gc) => gc.0.iter().flat_map(collect_lines).collect(),
        _ => vec![],
    }
}

/// Compute the length-weighted centroid of line segments.
/// Each segment's midpoint is weighted by its length.
fn compute_line_centroid(lines: &[&LineString<f64>]) -> Coord<f64> {
    let mut total_len = 0.0_f64;
    let mut cx = 0.0_f64;
    let mut cy = 0.0_f64;

    for line in lines {
        let coords: Vec<Coord<f64>> = line.coords().copied().collect();
        for i in 0..coords.len().saturating_sub(1) {
            let dx = coords[i + 1].x - coords[i].x;
            let dy = coords[i + 1].y - coords[i].y;
            let len = (dx * dx + dy * dy).sqrt();
            total_len += len;
            cx += len * (coords[i].x + coords[i + 1].x) / 2.0;
            cy += len * (coords[i].y + coords[i + 1].y) / 2.0;
        }
    }

    if total_len == 0.0 {
        // Degenerate: all zero-length segments — use first point
        let first_line = lines[0];
        return first_line.0[0];
    }

    Coord {
        x: cx / total_len,
        y: cy / total_len,
    }
}

fn distance_sq(a: Coord<f64>, b: Coord<f64>) -> f64 {
    let dx = a.x - b.x;
    let dy = a.y - b.y;
    dx * dx + dy * dy
}
